tracer/src/Datadog.Trace/Debugger/ExceptionAutoInstrumentation/ExceptionNormalizer.cs

```diff
@@ -0,0 +1,107 @@
+{
+    internal class ExceptionNormalizer
+    {
+        /// <summary>
+        /// Given the string representation of an exception alongside it's FQN of the outer and (potential) inner exception,
+        /// this function cleanse the stack trace from error messages, customized information attached to the exception and PDB line info if present.
+        /// It returns a hash representing the resulting cleansed exception and inner exceptions.
+        /// Used to aggregate same/similar exceptions that only differ by non-relevant bits.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal int NormalizeAndHashException(string exceptionString, string outerExceptionType, string? innerExceptionType)
+        {
+            if (string.IsNullOrEmpty(exceptionString))
+            {
+                throw new ArgumentException(@"Exception string cannot be null or empty", nameof(exceptionString));
+            }
+
+            var fnvHashCode = Fnv1aHash.FnvOffsetBias;
+
+            fnvHashCode = HashLine(VendoredMicrosoftCode.System.MemoryExtensions.AsSpan(outerExceptionType), fnvHashCode);
+
+            if (innerExceptionType != null)
+            {
+                fnvHashCode = HashLine(VendoredMicrosoftCode.System.MemoryExtensions.AsSpan(innerExceptionType), fnvHashCode);
+            }
+
+            var exceptionSpan = VendoredMicrosoftCode.System.MemoryExtensions.AsSpan(exceptionString);
+
+            var inSpan = VendoredMicrosoftCode.System.MemoryExtensions.AsSpan(" in ");
+
+            while (!exceptionSpan.IsEmpty)
+            {
+                var lineEndIndex = exceptionSpan.IndexOfAny('\r', '\n');
+                VendoredMicrosoftCode.System.ReadOnlySpan<char> line;
+
+                if (lineEndIndex >= 0)
+                {
+                    line = exceptionSpan.Slice(0, lineEndIndex);
+                    exceptionSpan = exceptionSpan.Slice(lineEndIndex + 1);
+                    if (!exceptionSpan.IsEmpty && exceptionSpan[0] == '\n')
+                    {
+                        exceptionSpan = exceptionSpan.Slice(1);
+                    }
+                }
+                else
+                {
+                    line = exceptionSpan;
+                    exceptionSpan = default;
+                }
+
+                if (IsStackTraceLine(line) && !IsLambdaFunctionLine(line))
+                {
+                    var index = VendoredMicrosoftCode.System.MemoryExtensions.IndexOf(line, inSpan);
+                    fnvHashCode = HashLine(index > 0 ? line.Slice(0, index) : line, fnvHashCode);
+                }
+            }
+
+            return fnvHashCode;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsStackTraceLine(VendoredMicrosoftCode.System.ReadOnlySpan<char> line)
+        {
+            int i = 0;
+
+            while (i < line.Length && char.IsWhiteSpace(line[i]))
```

Comment:
Is there a reason not doing here `line.TrimStart().StartsWith("at ", StringComparison.Ordinal)`?

Reply:
Not really, it was due to games I've done as part of benchmarking + performance tuning.
Improved in: d965ec2da2cc6da232af4e0f5222c4ad7fe00a33.